sstables/mx/writer.cc

```diff
@@ -1472,9 +1472,16 @@ void writer::consume_end_of_stream() {
             { large_data_type::elements_in_collection, std::move(_elements_in_collection_entry) },
         }
     });
-    const dht::sharder& sharder = _cfg.erm ? _cfg.erm->get_sharder(_schema)
-                                           : _schema.get_sharder(); // Used in tests
-    _sst.write_scylla_metadata(_shard, sharder, std::move(features), std::move(identifier), std::move(ld_stats), _cfg.origin);
+    {
+        locator::effective_replication_map_ptr erm;
+        if (_cfg.get_erm) {
+            erm = _cfg.get_erm();
+        }
+        const dht::sharder& sharder = erm ? erm->get_sharder(_schema)
+                                          : _schema.get_sharder(); // Used in tests
+        _sst.write_scylla_metadata(_shard, sharder, std::move(features), std::move(identifier), std::move(ld_stats),
+                                   _cfg.origin);
```

Comment:
I think there's a simpler option: don't write sharding info to write_scylla_metadata. Can revert 17d61635481376cdadcfda40f98cb7fdd019fc71 and make the sharder in config optional. We won't emit "this shard's ranges" metadata, but we don't need to, because the sstable will belong to only one shard, this tablet's shard, which will be inferred from first_key/last_key.

This removes the big layering violation moving cluster topology into the sstables layer.